stan/math/gpu/err/check_matrix_gpu.hpp

```diff
@@ -0,0 +1,222 @@
+      = opencl_context.get_kernel("check_symmetric");
+  cl::CommandQueue cmd_queue = opencl_context.queue();
+  cl::Context& ctx = opencl_context.context();
+  int symmetric_flag = 0;
+  try {
+    cl::Buffer buffer_symmetric_flag(ctx, CL_MEM_READ_WRITE, sizeof(int));
+
+    cmd_queue.enqueueWriteBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
+                                 &symmetric_flag);
+
+    kernel_check_symmetric.setArg(0, y.buffer());
+    kernel_check_symmetric.setArg(1, y.rows());
+    kernel_check_symmetric.setArg(2, y.cols());
+    kernel_check_symmetric.setArg(3, buffer_symmetric_flag);
+    kernel_check_symmetric.setArg(4, math::CONSTRAINT_TOLERANCE);
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_symmetric, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
+                                &symmetric_flag);
+  } catch (const cl::Error& e) {
+    check_opencl_error("symmetric_check", e);
+  }
+  //  if the matrix is not symmetric
+  if (symmetric_flag) {
+    domain_error(function, name, "is not symmetric", "");
+  }
+}
+/**
+ * Check if the specified matrix on the GPU has zeros on the diagonal
+ *
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y Matrix to test
+ * @param pos_def_check Test if matrix y is positive definite
+ *
+ * @throw <code>std::domain_error</code> if
+ *    any diagonal element of the matrix is zero.
+ */
+inline void check_diagonal_zeros(const char* function, const char* name,
+                                 const matrix_gpu& y,
+                                 const int pos_def_check = 0) {
+  if (y.size() == 0)
+    return;
+
+  cl::Kernel kernel_check_diagonal_zeros
+      = opencl_context.get_kernel("check_diagonal_zeros");
+  cl::CommandQueue cmd_queue = opencl_context.queue();
+  cl::Context ctx = opencl_context.context();
+  int flag = 0;
+  try {
+    cl::Buffer buffer_flag(ctx, CL_MEM_READ_WRITE, sizeof(int));
+
+    cmd_queue.enqueueWriteBuffer(buffer_flag, CL_TRUE, 0, sizeof(int), &flag);
+
+    kernel_check_diagonal_zeros.setArg(0, y.buffer());
+    kernel_check_diagonal_zeros.setArg(1, y.rows());
+    kernel_check_diagonal_zeros.setArg(2, y.cols());
+    kernel_check_diagonal_zeros.setArg(3, buffer_flag);
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_diagonal_zeros, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_flag, CL_TRUE, 0, sizeof(int), &flag);
+  } catch (const cl::Error& e) {
+    check_opencl_error("diag_zeros_check", e);
+  }
+  //  if zeros were found on the diagonal
+  if (flag) {
+    if (pos_def_check) {
+      domain_error(function, name, "is not positive definite", "");
```

Comment:
Does this do something to check for pos def vs. having zeroes on the diagonal?

Reply:
If not let's remove this because it doesn't follow the definition of checking for positive definite that the other pos_def check the stan math library has for cpus